### PR TITLE
High radix blocking Bcast algorithms

### PR DESCRIPTION
**Pull request by nusislam**
## Pull Request Description
This PR contains the kary and knomial tree-based Bcast algorithms. It also has a tree-based pipelined algorithm for large messages.

## Author Checklist
* [ ] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by nusislam** [comment]
test:mpich/ch4/ofi

File: src/mpi/coll/cvars.txt [hunk 1]
```diff
@@ -156,6 +156,61 @@ cvars:
...
+      default     : 2
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        k value for tree (kary, knomial, etc.) based bcast
+
+    - name        : MPIR_CVAR_BCAST_TREE_TYPE
+      category    : COLLECTIVE
+      type        : string
```
Comment (hzhou) [inline src/mpi/coll/cvars.txt:174]: I don't remember, but why these are not an enum?

File: src/mpi/coll/bcast/bcast_intra_pipelined_tree.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+/*
+ *
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ * Portions of this code were written by Intel Corporation.
+ * Copyright (C) 2011-2019 Intel Corporation. Intel provides this material
+ * to Argonne National Laboratory subject to Software Grant and Corporate
+ * Contributor License Agreement dated February 8, 2012.
+ */
```
Comment (hzhou) [inline src/mpi/coll/bcast/bcast_intra_pipelined_tree.c:11]: Old style, need fix.

File: src/mpi/coll/bcast/bcast_intra_pipelined_tree.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    if (is_contig) {    /* no need to pack */
+        MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+        sendbuf = (char *) buffer + true_lb;
+    } else {
+        MPIR_CHKLMEM_MALLOC(sendbuf, void *, nbytes, mpi_errno, "sendbuf", MPL_MEM_BUFFER);
+        if (rank == root) {
+            mpi_errno = MPIR_Typerep_pack(buffer, count, datatype, 0, sendbuf, nbytes,
+                                          &actual_packed_unpacked_bytes);
+            if (mpi_errno)
+                MPIR_ERR_POP(mpi_errno);
```
Comment (hzhou) [inline src/mpi/coll/bcast/bcast_intra_pipelined_tree.c:71]: use `MPIR_ERR_CHECK(mpi_errno);`

File: src/mpi/coll/bcast/bcast_intra_pipelined_tree.c [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    if (tree_type == MPIR_TREE_TYPE_KARY) {
+        lrank = (rank + (comm_size - root)) % comm_size;
+        parent = (lrank == 0) ? -1 : (((lrank - 1) / branching_factor) + root) % comm_size;
+        num_children = branching_factor;
+    } else {
+        mpi_errno =
+            MPIR_Treealgo_tree_create(rank, comm_size, tree_type, branching_factor, root, &my_tree);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
```
Comment (hzhou) [inline src/mpi/coll/bcast/bcast_intra_pipelined_tree.c:87]: same

File: src/mpi/coll/bcast/bcast_intra_pipelined_tree.c [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+                        MPIC_Irecv((char *) sendbuf + offset, msgsize, MPI_BYTE,
+                                   src, MPIR_BCAST_TAG, comm_ptr, &reqs[num_req++]);
+                    if (mpi_errno) {
+                        /* for communication errors, just record the error but continue */
+                        *errflag =
+                            MPIX_ERR_PROC_FAILED ==
+                            MPIR_ERR_GET_CLASS(mpi_errno) ? MPIR_ERR_PROC_FAILED : MPIR_ERR_OTHER;
+                        MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
+                        MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
+                    }
```
Comment (hzhou) [inline src/mpi/coll/bcast/bcast_intra_pipelined_tree.c:123]: use `MPIR_ERR_COLL_CHECKANDCONT`

File: src/mpi/coll/bcast/bcast_intra_pipelined_tree.c [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                    mpi_errno =
+                        MPIC_Irecv((char *) sendbuf + offset, msgsize, MPI_BYTE,
+                                   src, MPIR_BCAST_TAG, comm_ptr, &reqs[num_req++]);
+                    if (mpi_errno) {
+                        /* for communication errors, just record the error but continue */
+                        *errflag =
+                            MPIX_ERR_PROC_FAILED ==
+                            MPIR_ERR_GET_CLASS(mpi_errno) ? MPIR_ERR_PROC_FAILED : MPIR_ERR_OTHER;
+                        MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
+                        MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
```
Comment (hzhou) [inline src/mpi/coll/bcast/bcast_intra_pipelined_tree.c:143]: same ...

File: src/mpi/coll/cvars.txt (hunk 1 above)
Comment (nusislam) [reply src/mpi/coll/cvars.txt:174]: This CVAR lets users select a tree type for Bcast. The values are defined to be consistent with Ibcast.

**Comment by nusislam** [comment]
test:mpich/ch4/ofi

**Comment by nusislam** [comment]
Any more reviews?

File: src/mpi/coll/bcast/bcast_intra_tree.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
+ /*
```
Comment (yfguo) [inline src/mpi/coll/bcast/bcast_intra_tree.c:1]: Please remove the white space before the comment.

**Review by yfguo — CHANGES_REQUESTED** [review]
Other than the minor white space problem, looks good to me.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

File: src/mpi/coll/bcast/bcast_intra_tree.c (hunk 7 above)
Comment (nusislam) [reply src/mpi/coll/bcast/bcast_intra_tree.c:1]: Sorry I was occupied with other things. Will fix this tonight.

**Comment by nusislam** [comment]
test:mpich/ch4/ofi

**Comment by nusislam** [comment]
@yfguo - I have addressed your reviews.

**Review by yfguo — APPROVED** [review]

**Pull request merged**